### ISSUE TITLE
DrawString renders some fonts with black boxes?

### ISSUE THREAD

**Issue by Pyroglyph** [MonoGame/MonoGame#5227]
Steps to reproduce:
- Create new MonoGame Windows project (DirectX)
- Install any pixelly font (definitely works/breaks with [this one](http://dl.dafont.com/dl/?f=minecraftia))
- Add the following SpriteFont to your Content (use the .mgcb, of course)

```
<?xml version="1.0" encoding="utf-8"?>
<XnaContent xmlns:Graphics="Microsoft.Xna.Framework.Content.Pipeline.Graphics">
  <Asset Type="Graphics:FontDescription">
    <FontName>Minecraftia</FontName>
    <Size>6</Size>
    <Spacing>0</Spacing>
    <UseKerning>true</UseKerning>
    <Style>Regular</Style>
    <DefaultCharacter>?</DefaultCharacter>
    <CharacterRegions>
      <CharacterRegion>
        <Start>&#32;</Start>
        <End>&#126;</End>
      </CharacterRegion>
    </CharacterRegions>
  </Asset>
</XnaContent>
```

- Load the SpriteFont in Game1.cs (Content.Load<SpriteFont>)
- Begin the SpriteBatch with the SamplerState as PointClamp (so the font isn't blurred, still breaks without this though) and set the BlendState to anything because it breaks with any BlendState
- Draw the text in any color/position/size. I tested with this: `sb.DrawString(Font, "Text Rendering Problem", Vector2.Zero, Color.Yellow, 0f, Vector2.Zero, new Vector2(2, 2), SpriteEffects.None, 0f);`
- Note the black boxes around each letter and random black pixels below the letters themselves.

Result:
![image](https://cloud.githubusercontent.com/assets/9498104/19447728/c2a2364c-9496-11e6-91b5-c96750d8a5ee.png)

**Comment by theZMan** [comment]
Just to clarify you are alpha blending like this:

batch.Begin(SpriteSortMode.Texture,  **BlendState.AlphaBlend**, ....

and depending on the font it blends or does not?

If you use that font in one of the working XNA samples does it also repro?

**Comment by nkast** [comment]
> Note the black boxes around each letter and random black pixels below the letters themselves.

Can you check the return value of `Font.MeasureString("A");` ?

@theZMan 
>Just to clarify you are alpha blending like this:
>batch.Begin(SpriteSortMode.Texture, BlendState.AlphaBlend, ....

Parts of the font are already transparent...

**Comment by theZMan** [comment]
Actually they are not - DrawString actually drawn the minimal rectangle for each glyph. Your Screenshot is what  rendering looks like if alpha blending is disabled which is why I asked. I thought alpha was the default though - so using SpriteBatch.Begin() with no parameters should work too.

e.g. here is the XNA tutorial on fonts https://msdn.microsoft.com/en-us/library/bb447673.aspx

**Comment by nkast** [comment]
@theZMan That's what I though too at first. It seems there something else where.
What about those 2 dots? They are about 3 times the normal height bellow the text. They look like they appear below the space glyph.

**Comment by Pyroglyph** [comment]
@nkast e, P and R are all 6x11 and the space is 2x11
@theZMan If you mean the Kootenay font, it works with that. It works with most fonts, just not the pixel-perfect ones. See the image for a better description.
![image](https://cloud.githubusercontent.com/assets/9498104/19490626/cf6c2dca-9567-11e6-9270-8de65aac47e1.png)

**Comment by theZMan** [comment]
OK thats super wierd... the only other thing I can think of is that you changed the alpha properties of the .spritefont file in the content builder/pipeline? Usually the defaults work fine but maybe you changed them somehow. I don't have MonoGame installed here so I can't send a screen shot.

**Comment by Pyroglyph** [comment]
The four SpriteFont files I used for that screenshot are identical apart from the Font name. Something I also found out is that if you change the font size within the SpriteFont file, the text renders the alpha correctly (albeit the text is blurred since the size is not pixel-perfect).

**Comment by theZMan** [comment]
Even wierder... It this point you or someone is going to have to debug the content processor..

**Comment by Pyroglyph** [comment]
I tried to work around this today by setting all the black pixels in the SpriteFont object's Texture2D for transparent ones (using Color), but I can't use GetData with a Color array because the font texture is in the Dxt1 format. So I tried it with a byte array instead (yes it's the correct size), and found that each black pixel has a value of 0. That made me think "Hold on, if black is 0, then what is transparent?". After that I kinda just gave up. The only other solution I can think of is using a shader to remove the black on-the-fly but that would be an awful performance issue just to get the text working. I'm stuck.

**Comment by KonajuGames** [comment]
> I can't use GetData with a Color array because the font texture is in the Dxt1 format.

That's a valuable piece of information right there.  DXT1 does not support alpha, hence the black background in each glyph.  So when the texture was generated in the content pipeline, it had solid alpha in every pixel, so when it was compressed it chose DXT1 instead of DXT3 or DXT5 which do support alpha.  I'm guessing that when a font contains a bitmap for a specific small size, FreeType will rasterize the glyph with full alpha in every pixel, including the black pixels.

**Comment by theZMan** [comment]
Excellent detective work @KonajuGames 

XNA and MG use a heuristic to determine if they should use DXT1 and the heuristic is - are there any partial alpha  values.. For every other font in the world the edges don't align with the pixel grid so there is always partial alpha somewhere in the texture. For a blocky font at a size where the blocks match the pixel grid there is not.

See https://blogs.msdn.microsoft.com/shawnhar/2008/10/28/texture-compression/

DXT1 should be able to support alpha like this though - see https://msdn.microsoft.com/en-us/library/windows/desktop/bb147243(v=vs.85).aspx so seems like MonoGame isn't reading the textures properly.

The quick workaround is to change the font to use Color not DxtCompressed. That will encode the alpha properly. The texture files will get bigger unfortunately.

**Comment by KonajuGames** [comment]
DXT1 with 1-bit alpha is usually referred to as DXT1a, but I don't recall that OpenGL has support for it.  I'll have to look into it.

There is a special way of encoding DXT3 or DXT5 to avoid the block compression artifacts with high-contrast greyscale images such as fonts.  We will be looking at this soon as well.

> The texture files will get bigger unfortunately.

Generally only small point sizes are represented as bitmaps rather than vectors, so the increase in size won't be massive as the texture page should be quite small anyway.

**Comment by theZMan** [comment]
>DXT1 with 1-bit alpha is usually referred to as DXT1a, 
>but I don't recall that OpenGL has support for it. I'll have to look into it.
Looks likes its not obvious but is supported https://www.opengl.org/wiki/S3_Texture_Compression#DXT1_with_1-bit_Alpha

>Generally only small point sizes are represented as bitmaps rather than vectors, 
>so the increase in size won't be massive as the texture page should be quite small anyway.
 I meant the resultant XNB files - they will have a Color format not a DXT1 format so will be 8 times bigger.

**Comment by KonajuGames** [comment]
> Looks likes its not obvious but is supported

Cool.  Shouldn't be too difficult then.

**Comment by tomspilman** [comment]
>  it had solid alpha in every pixel, so when it was compressed it chose DXT1 instead of DXT3 or DXT5 which do support alpha.

In XNA fonts are always DXT3 regardless of the content.  It seems like maybe we should add support to override the decision made by the texture processor to force a gradient alpha format.

**Comment by Jjagg** [comment]
Fixed in #5398.
